### PR TITLE
Add HMM missing-observation support and an HMM missingness tutorial

### PR DESCRIPTION
**Pull request by mattlevine22**
This PR adds NaN-based missing-observation support to the exact HMM filter by reusing the shared masked observation log-probability machinery already used on the simulator side. Fully missing rows now act as neutral observation updates, and partial missingness is supported for exact-marginalizable HMM observation families such as MultivariateNormal and factorizable Independent(..., 1) observations. Internally, the prepared observation mask/safe-value views are computed once and reused across filtering and simulator paths rather than being recomputed downstream.

It also adds regression coverage for the supported and unsupported HMM missingness cases, updates the HMMConfig docs/checker behavior, and introduces a new Part 11c tutorial showing the two clean built-in HMM missingness workflows: multivariate Gaussian observations and independent categorical observations. The tutorial index is updated to link to the new notebook.

### CODE REVIEW

File: dynestyx/ops.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    _default_sample_interpretation = fn
+
+
+@defop
+def _sample_intp(
+    name: str,
+    dynamics: DynamicalModel,
+    *,
+    obs_times: Real[Array, "*obs_time_plate obs_time"] | None = None,
+    obs_values: Real[Array, "*obs_value_plate obs_time observation_dim"]
```
[the remark this answers is not in the document]
Comment (mattlevine22) [reply dynestyx/ops.py]: Ah shoot I was separately playing with default interpretations and that got mixed in here on accident at the end. Need to fix that before continuing

File: dynestyx/observation_missingness.py [hunk 2]
```diff
@@ -65,6 +65,206 @@ def _lift_scalar_observation_distribution(
     )
 
 
+def prepare_observation_mask(
+    obs_values: Float[Array, "time observation_dim"],
+) -> tuple[
+    Float[Array, "time observation_dim"],
+    Bool[Array, "time observation_dim"],
+    Bool[Array, " time"],
```
[the remark this answers is not in the document]
Comment (mattlevine22) [reply dynestyx/observation_missingness.py:187]: nope its used when there is 1 axis

File: dynestyx/observation_missingness.py [hunk 3]
```diff
@@ -87,9 +287,9 @@ class ObservationLogProb:
...
     obs_values: Float[Array, "time observation_dim"]
-    distribution_mode: Literal[
-        "uninitialized", "masked", "multivariate_normal", "independent"
-    ] = dataclasses.field(init=False, default="uninitialized")
+    precomputed_safe_obs: Array | None = None
+    precomputed_obs_mask: Bool[Array, "time observation_dim"] | None = None
+    distribution_mode: ObservationDistributionMode = dataclasses.field(init=False)
     safe_obs: Float[Array, "time observation_dim"] = dataclasses.field(init=False)
     obs_mask: Bool[Array, "time observation_dim"] = dataclasses.field(init=False)
     row_has_any_observed: Bool[Array, " time"] = dataclasses.field(init=False)
```
[the remark this answers is not in the document]
Comment (mattlevine22) [reply dynestyx/observation_missingness.py:418]: nope its used when there is 1 axis

File: dynestyx/observation_missingness.py [hunk 4]
```diff
@@ -65,6 +65,206 @@ def _lift_scalar_observation_distribution(
...
+        has_partial_missing,
+        has_fully_missing_rows,
+        observation_dim,
+    )
+
+
+def summarize_observation_mask(
+    obs_mask: Bool[Array, "time observation_dim"],
+) -> tuple[
+    Bool[Array, " time"],
```
[the remark this answers is not in the document]
Comment (mattlevine22) [reply dynestyx/observation_missingness.py:224]: nope its used when there is 1 axis

File: dynestyx/ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattlevine22) [reply dynestyx/ops.py]: fixed

File: docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb
Comment (DanWaxman) [inline docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb:1]: Leaving notebook comments here

Comment (DanWaxman) [reply docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb:1]: > # Part 11c: Missing Observations in HMMs

Should we just renumber these at some point? lol

Comment (DanWaxman) [reply docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb:1]: I have trouble reading these plots

Image

In particular, the lines are a bit confusing to me

Comment (DanWaxman) [reply docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb:1]: This is only a little concrete, in that it doesn't say anything about what the emission distributions actually are:

> Concretely, each coordinate is a 2-class categorical observation, and the full observation model is
> 
> $$
> p(y_t \mid z_t) = p(y_t^{(1)} \mid z_t)\, p(y_t^{(2)} \mid z_t).
> $$

File: dynestyx/inference/filters.py [hunk 5]
```diff
@@ -210,6 +210,9 @@ def _add_log_factors(
         obs_values: Real[Array, "*obs_value_plate obs_time observation_dim"]
         | Real[Array, "*obs_value_plate obs_time"]
         | None = None,
+        _obs_values_safe: Array | None = None,
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]: Not super clear what `_obs_values_safe` means, it should be documented

File: dynestyx/inference/filters.py
Comment (DanWaxman) [inline dynestyx/inference/filters.py:240]: Missing new args

File: dynestyx/inference/filters.py [hunk 6]
```diff
@@ -453,13 +485,11 @@ def compute_output_member(dyn, ot, ov, ct, cv, k, *idxs):
...
                     is_leaf=lambda x: x is None,
                 )
-                return compute_output(dyn, ot, ov, ct, cv, k)
+                return compute_output(dyn, ot, ov, ovs, om, ct, cv, k)
 
             idx_arrays = [jnp.arange(s) for s in plate_shapes]
             n_plates = len(plate_shapes)
             vmapped = compute_output_member
-            # Wrap w (innermost-first) maps plate dim d = n_plates - 1 - w, so the
-            # index array for that dim is mapped on axis 0 only at that level.
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py:462]: Would rather these not be removed

File: dynestyx/handlers.py [hunk 7]
```diff
@@ -2,24 +2,134 @@
...
+        "Categorical observations must be encoded as zero-based integer labels "
+        f"0..K-1 for the probed observation distribution with K={support_size}.",
+    )
+
+    safe_obs = jnp.where(
+        obs_mask,
+        obs_arr,
+        jnp.asarray(CATEGORICAL_MISSING_SENTINEL, dtype=obs_arr.dtype),
+    )
+    return safe_obs.astype(jnp.int32), obs_mask, has_missing
```
Comment (DanWaxman) [inline dynestyx/handlers.py]: I would rather these live somewhere else -- right now, handlers is pretty thin code + documentation, which I think is helpful

File: dynestyx/observation_missingness.py [hunk 8]
```diff
@@ -65,6 +65,206 @@ def _lift_scalar_observation_distribution(
...
+    bool,
+    bool,
+    bool,
+    int,
+]:
+    """Summarize row-wise missing-observation metadata from a boolean mask."""
```
Comment (DanWaxman) [inline dynestyx/observation_missingness.py]: More docstring on what is actually returned would be helpful

File: dynestyx/simulators.py
Comment (DanWaxman) [inline dynestyx/simulators.py:1]: I know it's not the fault of this PR, per se, but this file is getting wayy too long.

**Review by DanWaxman — CHANGES_REQUESTED** [review]

File: docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb
Comment (mattlevine22) [reply docs/tutorials/gentle_intro/11c_missing_observations_hmms.ipynb:1]: ok I updated the notebook to try help with these things. I'll fix the notebook names in a separate PR

File: dynestyx/inference/filters.py (hunk 5 above)
Comment (mattlevine22) [reply dynestyx/inference/filters.py]: its the filled version (i.e. with sentinels). added documentation and also changed the names to `filled` instead of `safe` for clarity.

File: dynestyx/inference/filters.py
Comment (mattlevine22) [reply dynestyx/inference/filters.py:240]: added thanks.

File: dynestyx/inference/filters.py (hunk 6 above)
Comment (mattlevine22) [reply dynestyx/inference/filters.py:462]: un-undid

File: dynestyx/handlers.py (hunk 7 above)
Comment (mattlevine22) [reply dynestyx/handlers.py]: okay moved them to observation_missingness.py.

File: dynestyx/observation_missingness.py (hunk 8 above)
Comment (mattlevine22) [reply dynestyx/observation_missingness.py]: added docstring

File: dynestyx/simulators.py
Comment (mattlevine22) [reply dynestyx/simulators.py:1]: yeah but let's do it after? was thinking better to wait for Theo but we can discuss that as well.

File: dynestyx/inference/filters.py [hunk 9]
```diff
@@ -342,7 +361,7 @@ def _add_log_factors_batched(
                 )
             output_kind = "continuous"
 
-            def compute_output(dyn, ot, ov, ct, cv, k):
+            def compute_output(dyn, ot, ov, ovs, om, ct, cv, k):
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
            def compute_output(dyn, ot, ov, ovf, om, ct, cv, k):
```

File: dynestyx/inference/filters.py [hunk 10]
```diff
@@ -355,20 +374,23 @@ def compute_output(dyn, ot, ov, ct, cv, k):
 
         elif isinstance(config, HMMConfigs):
             output_kind = "hmm"
+            uses_preprocessed_obs = True
 
-            def compute_output(dyn, ot, ov, ct, cv, k):
+            def compute_output(dyn, ot, ov, ovs, om, ct, cv, k):
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
            def compute_output(dyn, ot, ov, ovf, om, ct, cv, k):
```

File: dynestyx/inference/filters.py [hunk 11]
```diff
@@ -355,20 +374,23 @@ def compute_output(dyn, ot, ov, ct, cv, k):
...
                 return compute_hmm_filter(
                     dyn,
                     obs_times=ot,
                     obs_values=ov,
+                    _obs_values_filled=ovs,
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
                    _obs_values_filled=ovf,
```

File: dynestyx/inference/filters.py [hunk 12]
```diff
@@ -355,20 +374,23 @@ def compute_output(dyn, ot, ov, ct, cv, k):
...
+                    _obs_mask=om,
                     ctrl_values=cv,
                 )
 
         elif isinstance(config, DiscreteTimeConfigs):
             if config.filter_source == "cuthbert":
                 output_kind = "cuthbert"
 
-                def compute_output(dyn, ot, ov, ct, cv, k):
+                def compute_output(dyn, ot, ov, ovs, om, ct, cv, k):
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
                def compute_output(dyn, ot, ov, ovf, om, ct, cv, k):
```

File: dynestyx/inference/filters.py [hunk 13]
```diff
@@ -382,7 +404,7 @@ def compute_output(dyn, ot, ov, ct, cv, k):
             elif config.filter_source == "cd_dynamax":
                 output_kind = "cd_dynamax_discrete"
 
-                def compute_output(dyn, ot, ov, ct, cv, k):
+                def compute_output(dyn, ot, ov, ovs, om, ct, cv, k):
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
                def compute_output(dyn, ot, ov, ovf, om, ct, cv, k):
```

File: dynestyx/inference/filters.py [hunk 14]
```diff
@@ -426,7 +448,24 @@ def compute_output(dyn, ot, ov, ct, cv, k):
             ctrl_values=ctrl_values,
         )
         k_axis = 0 if keys is not None else None
-        base_axes = (dyn_axes, ot_axis, ov_axis, ct_axis, cv_axis, k_axis)
+        if uses_preprocessed_obs:
+            ovs_axis = _array_plate_axis(_obs_values_filled, plate_shapes)
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
            ovf_axis = _array_plate_axis(_obs_values_filled, plate_shapes)
```

File: dynestyx/inference/filters.py [hunk 15]
```diff
@@ -426,7 +448,24 @@ def compute_output(dyn, ot, ov, ct, cv, k):
...
+            om_axis = _array_plate_axis(_obs_mask, plate_shapes)
+        else:
+            ovs_axis = None
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
            ovf_axis = None
```

File: dynestyx/inference/filters.py [hunk 16]
```diff
@@ -426,7 +448,24 @@ def compute_output(dyn, ot, ov, ct, cv, k):
...
+            om_axis = None
+            _obs_values_filled = None
+            _obs_mask = None
+        base_axes = (
+            dyn_axes,
+            ot_axis,
+            ov_axis,
+            ovs_axis,
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
            ovf_axis,
```

File: dynestyx/inference/filters.py [hunk 17]
```diff
@@ -443,7 +482,7 @@ def compute_output(dyn, ot, ov, ct, cv, k):
         if ic_batched:
             orig_ic = dynamics.initial_condition
 
-            def compute_output_member(dyn, ot, ov, ct, cv, k, *idxs):
+            def compute_output_member(dyn, ot, ov, ovs, om, ct, cv, k, *idxs):
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
            def compute_output_member(dyn, ot, ov, ovf, om, ct, cv, k, *idxs):
```

File: dynestyx/inference/filters.py [hunk 18]
```diff
@@ -453,21 +492,25 @@ def compute_output_member(dyn, ot, ov, ct, cv, k, *idxs):
                     member_ic,
                     is_leaf=lambda x: x is None,
                 )
-                return compute_output(dyn, ot, ov, ct, cv, k)
+                return compute_output(dyn, ot, ov, ovs, om, ct, cv, k)
```
Comment (DanWaxman) [inline dynestyx/inference/filters.py]:
```suggestion
                return compute_output(dyn, ot, ov, ovf, om, ct, cv, k)
```

**Review by DanWaxman — APPROVED** [review]
Just some renaming comments. Otherwise LGTM!
